NitroxClient/Communication/ClientBridge.cs

```diff
@@ -0,0 +1,249 @@
+        }
+
+        public void ClaimReservation()
+        {
+            try
+            {
+                if (CurrentState != ClientBridgeState.Reserved)
+                {
+                    throw new ProhibitedClaimAttemptException();
+                }
+
+                var packet = new ClaimPlayerSlotReservation(correlationId, ReservationKey);
+                serverClient.Send(packet);
+
+                CurrentState = ClientBridgeState.Connected;
+            }
+            catch (Exception ex)
+            {
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+        }
+
+        public void HandleRejectedReservation(string correlationId, ReservationRejectionReason rejectionReason)
+        {
+            if (CurrentState == ClientBridgeState.Disconnected)
+            {
+                return;
+            }
+
+            try
+            {
+                if (CurrentState != ClientBridgeState.WaitingForRerservation)
+                {
+                    throw new ProhibitedReservationRejectionException();
+                }
+
+                ValidateCorrelationId(correlationId);
+                ConfirmCorrelationId(correlationId);
+                ValidateRejectionReason(rejectionReason);
+
+                ReservationKey = null;
+                ReservationRejectionReason = rejectionReason;
+
+                CurrentState = ClientBridgeState.ReservationRejected;
+            }
+            catch (Exception ex)
+            {
+                TryStopClient();
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+        }
+
+        public void Send(Packet packet)
+        {
+            if (CurrentState != ClientBridgeState.Connected)
+            {
+                throw new ProhibitedSendAttemptException();
+            }
+
+            serverClient.Send(packet);
+        }
+
+        public PacketSuppression<T> Suppress<T>()
+        {
+            return new PacketSuppression<T>(suppressedPacketsTypes);
+        }
+
+        private void TryStopClient()
+        {
+            if (serverClient.IsConnected)
+            {
+                serverClient.Stop();
+            }
+        }
+
+        private void RequestReservation(string playerName)
+        {
+            if (CurrentState == ClientBridgeState.WaitingForRerservation ||
+                CurrentState == ClientBridgeState.Reserved ||
+                CurrentState == ClientBridgeState.Connected)
+            {
+                throw new ProhibitedConnectAttemptException();
+            }
+
+            var reservePlayerSlot = new ReservePlayerSlot(correlationId, playerName);
+            serverClient.Send(reservePlayerSlot);
+
+            PlayerId = playerName;
+            CurrentState = ClientBridgeState.WaitingForRerservation;
+        }
+
+        private void ValidateCorrelationId(string correlationId)
+        {
+            if (correlationId == null)
```

Comment:
Will do, but I may be a little discerning in move the exception types. Some of them are only raised by the client bridge - which only exists in the client.

Reply:
And that is exactly why I mentioned to make the communication ones subclasses of the global ones, so that they can be dealt with separately when need arises. Though then it's a little hard to use `Validate`.